base/shell/explorer/desktop.cpp

```diff
@@ -22,107 +22,160 @@
-            to make the main thread wake up! */
-            return 1;
-        }
+    HRESULT Initialize(ITrayWindow* pTray);
+    void Destroy();
+};
 
-        _SHDesktopMessageLoop(hDesktop);
+/*******************************************************************/
 
-        /* FIXME: Properly rundown the main thread! */
-        ExitProcess(0);
+CDesktopThread::CDesktopThread()
+{
+    m_Tray = NULL;
+    m_hEvent = NULL;
+    m_hThread = NULL;
+}
 
-        return 0;
+CDesktopThread::~CDesktopThread()
+{
+    if (m_hThread)
+    {
+        CloseHandle(m_hThread);
     }
 
-    static DWORD CALLBACK s_DesktopThreadProc(IN OUT LPVOID lpParameter)
+    if (m_hEvent)
     {
-        return reinterpret_cast<CDesktopThread*>(lpParameter)->DesktopThreadProc();
+        CloseHandle(m_hEvent);
     }
+}
 
-public:
-    CDesktopThread() :
-        m_hEvent(NULL),
-        m_Tray(NULL)
+HRESULT CDesktopThread::Initialize(ITrayWindow* pTray)
+{
+    HANDLE Handles[2];
+
+    m_Tray = pTray;
+
+    m_hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
+
+    if (!m_hEvent)
     {
+        return E_FAIL;
     }
 
-    HRESULT Initialize(IN OUT ITrayWindow *pTray)
-    {
-        HANDLE hThread;
-        HANDLE Handles[2];
+    m_hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (LPVOID)this, 0, NULL);
+
+    if (!m_hThread)
+    {   
+        return E_FAIL;
+    }
+
+    Handles[0] = m_hThread;
+    Handles[1] = m_hEvent;
 
-        m_Tray = pTray;
+    while (true)
+    {
+        DWORD WaitResult = MsgWaitForMultipleObjects(_countof(Handles), Handles, FALSE, INFINITE, QS_ALLEVENTS);
 
-        m_hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
-        if (!m_hEvent)
-            return E_FAIL;
+        if (WaitResult == WAIT_OBJECT_0 + _countof(Handles))
+        {
+            TrayProcessMessages(m_Tray);
+        }
 
-        hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (PVOID)this, 0, NULL);
-        if (!hThread)
+        else if (WaitResult != WAIT_FAILED && WaitResult != WAIT_OBJECT_0)
         {
-            CloseHandle(m_hEvent);
-            return E_FAIL;
+            break;
         }
+    }
 
-        Handles[0] = hThread;
-        Handles[1] = m_hEvent;
+    return S_OK;
+}
+
+void CDesktopThread::Destroy()
+{
+    if (m_hThread)
+    {
+        DWORD dwExitCode;
 
-        for (;;)
+        GetExitCodeThread(m_hThread, &dwExitCode);
+
+        if (dwExitCode == STILL_ACTIVE)
         {
-            DWORD WaitResult = MsgWaitForMultipleObjects(_countof(Handles), Handles, FALSE, INFINITE, QS_ALLEVENTS);
-            if (WaitResult == WAIT_OBJECT_0 + _countof(Handles))
-            {
-                TrayProcessMessages(m_Tray);
-            }
-            else if (WaitResult != WAIT_FAILED && WaitResult != WAIT_OBJECT_0)
-            {
-                break;
-            }
+            TerminateThread(m_hThread, 0);
         }
+    }
+}
 
-        CloseHandle(hThread);
-        CloseHandle(m_hEvent);
```

Comment:
You're sure you don't need to close the handles?

Reply:
I moved them to the destructor.